internal/datacoord/import_checker.go

```diff
@@ -0,0 +1,394 @@
+			return
+		}
+	}
+}
+
+func (c *importChecker) checkFailure(jobID int64) {
+	tasks := c.imeta.GetTaskBy(WithJob(jobID))
+	var (
+		isFailed   bool
+		failedTask ImportTask
+	)
+	for _, task := range tasks {
+		if task.GetState() == internalpb.ImportState_Failed {
+			isFailed = true
+			failedTask = task
+			log.Warn("Import has failed, all tasks with the same jobID will be marked as failed", WrapTaskLog(task)...)
+			break
+		}
+	}
+	if isFailed {
+		for _, task := range tasks {
+			if task.GetState() == internalpb.ImportState_Failed {
+				continue
+			}
+			err := c.imeta.UpdateTask(task.GetTaskID(), UpdateState(internalpb.ImportState_Failed),
+				UpdateReason(failedTask.GetReason()))
+			if err != nil {
+				continue
+			}
+		}
+	}
+}
+
+func (c *importChecker) checkTimeout(jobID int64) {
+	tasks := c.imeta.GetTaskBy(WithStates(internalpb.ImportState_InProgress), WithJob(jobID))
+	if len(tasks) == 0 {
+		return
+	}
+	for _, task := range tasks {
+		if time.Since(task.GetLastActiveTime()) > 10*time.Minute {
+			log.Warn("task progress is stagnant", WrapTaskLog(task)...)
+		}
+	}
+
+	timeoutTime := tsoutil.PhysicalTime(c.imeta.GetJob(jobID).GetTimeoutTs())
+	if time.Now().After(timeoutTime) {
+		log.Warn("Import timeout, expired the specified time limit",
+			zap.Int64("jobID", jobID), zap.Time("timeoutTime", timeoutTime))
+		for _, task := range tasks {
+			err := c.imeta.UpdateTask(task.GetTaskID(), UpdateState(internalpb.ImportState_Failed),
+				UpdateReason("import timeout"))
+			if err != nil {
+				log.Warn("update task state failed", WrapTaskLog(task, zap.Error(err))...)
+			}
+		}
+	}
+}
+
+func (c *importChecker) checkCollection(collectionID int64) {
+	tasks := c.imeta.GetTaskBy(WithStates(internalpb.ImportState_Pending,
+		internalpb.ImportState_InProgress), WithCollection(collectionID))
+	if len(tasks) == 0 {
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	has, err := c.broker.HasCollection(ctx, collectionID)
+	if err != nil {
+		log.Warn("verify existence of collection failed", zap.Int64("collection", collectionID), zap.Error(err))
+		return
+	}
+	if !has {
+		for _, task := range tasks {
+			err = c.imeta.UpdateTask(task.GetTaskID(), UpdateState(internalpb.ImportState_Failed),
+				UpdateReason(fmt.Sprintf("collection %d dropped", collectionID)))
+			if err != nil {
+				log.Warn("update task state failed", WrapTaskLog(task, zap.Error(err))...)
+			}
+		}
+	}
+}
+
+func (c *importChecker) checkGC(jobID int64) {
+	GCRetention := Params.DataCoordCfg.ImportTaskRetention.GetAsDuration(time.Second)
+	tasks := c.imeta.GetTaskBy(WithStates(internalpb.ImportState_Failed, internalpb.ImportState_Completed), WithJob(jobID))
+	sort.Slice(tasks, func(i, j int) bool {
+		return tasks[i].GetLastActiveTime().After(tasks[j].GetLastActiveTime())
+	})
+	if len(tasks) == 0 {
+		return
+	}
+	if time.Since(tasks[0].GetLastActiveTime()) >= GCRetention {
```

Comment:
we should update LastActiveTime once the task state changed to ImportState_Failed or ImportState_Completed